### PR TITLE
implement user state, connect to apollo server

### LINKED ISSUE #2: Connect Login and Register forms to API
Use Apollo server for Login and Register data

### PR DESCRIPTION
**Pull request by Aaron-Goldman**
- Added global user state with sign in/sign out/register functions.
- Login form checks credentials against server query.
- Register form creates new user with server mutation.
- User state persists between sessions in local storage.

close #2

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -22,24 +23,26 @@ import {
...
+            />
+            <Route path={LOGIN_PATH}>
+              <Login />
+            </Route>
+            <Route path={REGISTER_PATH}>
+              <Register />
+            </Route>
+            <Route path={NOT_FOUND_PATH}>
+              <NotFound />
+            </Route>
```
Comment (ahmed-gz) [inline src/App.js]: For consistency I recommend using the same pattern you used in `<PrivateRoute>` which is also a lot shorter:
```
<Route path={LOGIN_PATH} component={Login} />
<Route path={REGISTER_PATH} component={Register} />
<Route path={NOT_FOUND_PATH} component={NotFound} />
```

File: src/Login/index.js [hunk 2]
```diff
@@ -15,7 +18,11 @@ function Login() {
   const handlePasswordInput = (event) => {
     setPassword(event.target.value);
   };
-  const handleSubmit = () => {};
+  const handleSubmit = async (event) => {
+    event.preventDefault();
+    await auth.signIn(username, password);
+    history.push('/');
```
Comment (ahmed-gz) [inline src/Login/index.js]: Let's use `HOME_PATH` here for consistency, it will be easier to do search/replace later if we want to change it

File: src/PrivateRoute.js [hunk 3]
```diff
@@ -2,28 +2,26 @@ import React from 'react';
...
     exact,
     component,
-    loggedIn,
   } = props;
-  const location = useLocation();
+  const auth = useAuth();
 
   return (
-    loggedIn ? (
+    auth.user ? (
```
Comment (ahmed-gz) [inline src/PrivateRoute.js]: It's always best to convert an object to a `Boolean` before using it in a condition like this: `!!auth.user`

File: src/PrivateRoute.js [hunk 4]
```diff
@@ -34,14 +32,12 @@ PrivateRoute.defaultProps = {
   path: '/',
```
Comment (ahmed-gz) [inline src/PrivateRoute.js]: `HOME_PATH` here

File: src/Register/index.js [hunk 5]
```diff
@@ -20,7 +24,11 @@ function Register() {
   const handlePasswordInput = (event) => {
     setPassword(event.target.value);
   };
-  const handleSubmit = () => {};
+  const handleSubmit = async (event) => {
+    event.preventDefault();
+    await auth.register(username, firstName, lastName, password);
+    history.push('/');
```
Comment (ahmed-gz) [inline src/Register/index.js]: `HOME_PATH` here

File: src/use-auth.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    user,
+    signIn,
+    signOut,
+    register,
+  };
+}
+
+export function ProvideAuth({ children }) {
+  const auth = useProvideAuth();
+  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
```
Comment (ahmed-gz) [inline src/use-auth.js]: Please use capital letter because it should look like a component: `AuthContext`

File: src/use-auth.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+  return {
+    user,
+    signIn,
+    signOut,
+    register,
+  };
+}
+
+export function ProvideAuth({ children }) {
```
Comment (ahmed-gz) [inline src/use-auth.js]: The standard pattern for providers is `FeatureProvider`, so it's best to go with `AuthProvider` instead

File: src/use-auth.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  const auth = useProvideAuth();
+  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
+}
+ProvideAuth.defaultProps = { children: <div /> };
+ProvideAuth.propTypes = { children: PropTypes.element };
+
+export default useAuth;
```
Comment (ahmed-gz) [inline src/use-auth.js]: `AuthProvider` should be the default export instead because it's the entry point to the `auth` module

File: src/use-auth.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    username
+    id
+    password
+  }
+}
+`;
+
+const authContext = createContext();
+
+const useAuth = () => useContext(authContext);
```
Comment (ahmed-gz) [inline src/use-auth.js]: `useAuth` can be a named export since it's a consuming hook, usually it's best to expose the main object or component as the default export and expose the utility like functions as named exports

File: src/use-auth.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
+import React, {
+  useState,
+  useContext,
+  createContext,
+} from 'react';
+import PropTypes from 'prop-types';
+import { useQuery, useMutation, gql } from '@apollo/client';
+
+const REGISTER = gql`
+  mutation RegisterMutation(
```
Comment (ahmed-gz) [inline src/use-auth.js]: let's pull mutations into a separate file `src/graphql/mutations.js`

File: src/use-auth.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+      firstName: $firstName
+      lastName: $lastName
+      password: $password
+      userRoleId: 3
+    ) {
+      id
+    }
+  }
+`;
+const QUERY_USERS = gql`
```
Comment (ahmed-gz) [inline src/use-auth.js]: let's pull queries into a separate file `src/graphql/queries.js`

File: src/use-auth.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    password
+  }
+}
+`;
+
+const authContext = createContext();
+
+const useAuth = () => useContext(authContext);
+
+function useProvideAuth() {
```
Comment (ahmed-gz) [inline src/use-auth.js]: Do you think you could use this hook anywhere else other than `AuthProvider`? if not then we can just move all the code here to the main provider and rename this file to `src/services/AuthProvider.js`

File: src/use-auth.js [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+  const { refetch } = useQuery(QUERY_USERS);
+  const [addUser] = useMutation(REGISTER);
+
+  const signIn = async (username, password) => {
+    const result = await refetch();
+    if (!result) return;
+    const match = result.data.users.find((element) => element.username === username);
+    if (!match || match.password !== password) return;
+    setUser(match.id);
+    localStorage.setItem('user', user);
```
Comment (ahmed-gz) [inline src/use-auth.js]: Be careful with optimistic updates because `user` might not always contain the most up to date value updated from `setUser`, so I would use match.id here instead

File: src/use-auth.js [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+function useProvideAuth() {
+  const [user, setUser] = useState(localStorage.getItem('user') || 0);
+  const { refetch } = useQuery(QUERY_USERS);
+  const [addUser] = useMutation(REGISTER);
+
+  const signIn = async (username, password) => {
+    const result = await refetch();
+    if (!result) return;
+    const match = result.data.users.find((element) => element.username === username);
+    if (!match || match.password !== password) return;
```
Comment (ahmed-gz) [inline src/use-auth.js]: I would suggest returning some error messages or codes instead of returning nothing, unless this is just temporary!

File: src/use-auth.js [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    localStorage.setItem('user', user);
+  };
+
+  const signOut = () => {
+    setUser(0);
+    localStorage.setItem('user', 0);
+  };
+
+  const register = async (username, firstName, lastName, password) => {
+    await addUser({
```
Comment (ahmed-gz) [inline src/use-auth.js]: I would suggest always returning something, like the response from `addUser`, because sometimes this can cause a race condition when you call register like this:
```
    await auth.register(username, firstName, lastName, password);
    history.push('/'); // it's possible in some case to redirect before the register finishes even with await
```

File: src/use-auth.js [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  }
+}
+`;
+
+const authContext = createContext();
+
+const useAuth = () => useContext(authContext);
+
+function useProvideAuth() {
+  const [user, setUser] = useState(localStorage.getItem('user') || 0);
```
Comment (ahmed-gz) [inline src/use-auth.js]: `user` here doesn't suggest what it's for exactly and it's hard to read its structure as well, I would suggest to be more descriptive because it feels like an object, so maybe you could go with: `loggedUserId` or `currentUserId`,...

**Review by ahmed-gz — CHANGES_REQUESTED** [review]
I really like the way you're thinking with and the pattern you used here, just needs a bit more polishing, left some suggestions 👍

File: src/PrivateRoute.js (hunk 3 above)
Comment (Aaron-Goldman) [reply src/PrivateRoute.js]: There's apparently an ESLint rule that disallows casting to Boolean when using an object directly as part of a condition, because the value is always cast to a Boolean anyway. https://eslint.org/docs/rules/no-extra-boolean-cast
This code is considered fine:
```
  const loggedIn = !!auth.user;

  return (
    loggedIn ? (
```
But this isn't:
```
  return (
    !!auth.user ? (
```
Will use the first one.

Comment (ahmed-gz) [reply src/PrivateRoute.js]: Ah ok, I didn't know that rule existed, your alternative looks much better because it's also easier to read. It might look obvious like this but if you have 3 or more conditions like that, code will be very hard to read without meaningful condition names, I would even go one step further and rename it `isLoggedIn` 

It's always best to name boolean variables in a name that sounds like asking a question that can only be answered with true or false, i.e. `isSomething`, `hasSomething`, `shouldSomething`,...

Comment (ahmed-gz) [reply src/PrivateRoute.js]: Here are some guides and articles that could be useful in this topic:
https://github.com/airbnb/javascript#naming-conventions
https://www.codereadability.com/refactoring-if-statements-for-readability/

File: src/use-auth.js [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+  const { refetch } = useQuery(QUERY_USERS);
+  const [addUser] = useMutation(REGISTER);
+
+  const signIn = async (username, password) => {
+    const result = await refetch();
+    if (!result) return;
+    const match = result.data.users.find((element) => element.username === username);
```
Comment (ahmed-gz) [inline src/use-auth.js]: For one liner array functions like map, find, filter,.. you can use one letter variable name to shorten the code a bit since we don't need a meaningful name for it in this case, the condition can still be self explanatory like this:
`result.data.users.find((u) => u.username === username)`

File: src/use-auth.js (hunk 12 above)
Comment (Aaron-Goldman) [reply src/use-auth.js]: I used the design pattern shown here: https://usehooks.com/useAuth/
If there was a reason for making a separate hook and provider in that example, I'm not sure what it was. If it was only needed there because it's storing the user state on the server side, which we aren't doing here, then I can change it.

Comment (ahmed-gz) [reply src/use-auth.js]: If you feel it's best to leave those methods in that hook I'm fine with that, we can always refactor it later if we need to.
Let's just rename the provider and context to what's more commonly used(I left some suggestions about that), here is an article that could be helpful about how to restructure that layer of the code: https://kentcdodds.com/blog/application-state-management-with-react
Feel free to go with what makes the most sense to you, we can always improve it once you finish all the other pieces. If you want to have a call to discuss it more in depth, just let me know 👍

File: src/use-auth.js [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+import { useQuery, useMutation } from '@apollo/client';
+import QUERY_USERS from './graphql/queries';
+import REGISTER from './graphql/mutations';
+
+const AuthContext = createContext();
+
+const useAuth = () => useContext(AuthContext);
+
+function useProvideAuth() {
+  const [user, setUser] = useState(localStorage.getItem('user') || undefined);
```
Comment (ahmed-gz) [inline src/use-auth.js]: @Aaron-Goldman `user` here doesn't suggest what it's for exactly and it's hard to read its structure as well, I would suggest to be more descriptive because it feels like an object, so maybe you could go with: `loggedUserId` or `currentUserId`,...

File: src/use-auth.js [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+  const { refetch: refetchUsers } = useQuery(QUERY_USERS);
+  const [addUser] = useMutation(REGISTER);
+
+  const signIn = async (username, password) => {
+    if (!username) throw Error('Username required.');
+    const result = await refetchUsers();
+    if (!result) throw Error('Server did not respond.');
+    const match = result.data.users.find((u) => u.username === username);
+    if (!match || match.password !== password) throw Error('Incorrect username or password.');
```
Comment (ahmed-gz) [inline src/use-auth.js]: You could move that condition in the previous line:
`result.data.users.find((u) => u.username === username && u.password === password);`

File: src/use-auth.js [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+const useAuth = () => useContext(AuthContext);
+
+function useProvideAuth() {
+  const [user, setUser] = useState(localStorage.getItem('user') || undefined);
+  const { refetch: refetchUsers } = useQuery(QUERY_USERS);
+  const [addUser] = useMutation(REGISTER);
+
+  const signIn = async (username, password) => {
+    if (!username) throw Error('Username required.');
```
Comment (ahmed-gz) [inline src/use-auth.js]: Shouldn't we have a similar one for password? 
`if (!password) throw Error('Password required.');`

File: src/Login/index.js [hunk 21]
```diff
@@ -1,21 +1,34 @@
...
   const [username, setUsername] = useState('');
   const [password, setPassword] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
 
   const handleUsernameInput = (event) => {
     setUsername(event.target.value);
+    setErrorMessage('');
   };
   const handlePasswordInput = (event) => {
     setPassword(event.target.value);
```
Comment (ahmed-gz) [inline src/Login/index.js:27]: Shouldn't we reset `errorMessage` here as well: `setErrorMessage('')`

File: src/use-auth.test.js [hunk 22]
```diff
@@ -0,0 +1,120 @@
+import React, { useState } from 'react';
+import { fireEvent, render, screen } from '@testing-library/react';
+import { MockedProvider } from '@apollo/client/testing';
+import { isApolloError } from '@apollo/client';
+import useAuth, { AuthProvider } from './use-auth';
+import QUERY_USERS from './graphql/queries';
+import REGISTER from './graphql/mutations';
+
+function TestComponent() {
```
Comment (ahmed-gz) [inline src/use-auth.test.js]: You shouldn't need to create a component to test the hook, you should test the existing pages that already consume that hook.
So basically you need to move those tests to the login and register pages instead: `src/Login/index.test.js` and `src/Register/index.test.js`

File: src/use-auth.js [hunk 23]
```diff
@@ -0,0 +1,68 @@
+import React, {
```
Comment (ahmed-gz) [inline src/use-auth.js]: Please rename this file to `AuthProvider.js`, the main feature for this file should be the `AuthProvider` not the hook.
And `AuthProvider` should be the default export, it should be the entry point of this file, hooks are just functions to consume the provider.

**Review by ahmed-gz — CHANGES_REQUESTED** [review]
Looks good overall, the major work needed is moving the tests from the `AuthProvider` to the components that consume it.

File: src/AuthProvider.js [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+function useProvideAuth() {
+  const [user, setUser] = useState(localStorage.getItem('user') || undefined);
+  const { refetch: refetchUsers } = useQuery(QUERY_USERS);
+  const [addUser] = useMutation(REGISTER);
+
+  const signIn = async (username, password) => {
+    const result = await refetchUsers();
+    if (!result) throw Error('Server did not respond.');
+    const match = result.data.users.find((u) => u.username === username);
+    if (!match || match.password !== password) throw Error('Incorrect username or password.');
```
Comment (ahmed-gz) [inline src/AuthProvider.js]: You could move that condition to the previous line:
`result.data.users.find((u) => u.username === username && u.password === password);`

File: src/Login/index.test.js [hunk 25]
```diff
@@ -1,15 +1,64 @@
...
+      </AuthProvider>
+    </MockedProvider>,
+  );
   expect(screen.getByRole('form')).toBeInTheDocument();
-  expect(screen.getByText('Register')).toBeInTheDocument();
-  userEvent.type(screen.getByLabelText('Username'), 'User');
-  expect(screen.getByDisplayValue('User')).toBeInTheDocument();
-  userEvent.type(screen.getByLabelText('Password'), 'password');
+  expect(screen.getByText('Register Now')).toBeInTheDocument();
+  const loginButton = screen.getAllByRole('button')[0];
```
Comment (ahmed-gz) [inline src/Login/index.test.js]: This looks good, you can also use destructing here if you'd like:
`const [loginButton] = screen.getAllByRole('button');`

File: src/Login/index.test.js [hunk 26]
```diff
@@ -1,15 +1,64 @@
...
+  userEvent.type(passwordField, 'password');
   expect(screen.getByDisplayValue('password')).toBeInTheDocument();
+
+  expect(loginButton).toBeEnabled();
+
+  fireEvent.click(loginButton);
+  expect(await screen.findByText('Incorrect username or password.')).toBeInTheDocument();
+  userEvent.type(usernameField, '{backspace}');
+  fireEvent.click(loginButton);
+  expect(screen.queryByText('Incorrect username or password.')).not.toBeInTheDocument();
```
Comment (ahmed-gz) [inline src/Login/index.test.js]: I think we should also check that the `localStorage` has the user value after the login is successful or doesn't have it when unsuccessful

**Review by ahmed-gz — CHANGES_REQUESTED** [review]
You're almost there, looks good overall, left some suggestions 👍

File: src/Login/index.test.js [hunk 27]
```diff
@@ -1,15 +1,75 @@
...
+        },
+      ],
+    },
+  },
+};
+
+const mocks = [
+  mockedUsersResponse,
+  mockedUsersResponse,
+  mockedUsersResponse,
```
Comment (ahmed-gz) [inline src/Login/index.test.js]: I'm not sure why you needed to duplicate `mockedUsersResponse` here 3 times, but if it's to have 3 users in the array then you should add 2 more users with different properties in `mockedUsersResponse`, not sure if it would work like this but let me know if you have issues with it, ideally we want to have something more realistic like this:
```
      users: [
        {
          username: 'Danny',
          password: '632569',
          id: 1,
        },
        {
          username: 'James',
          password: '12345',
          id: 2,
        },
        {
          username: 'William',
          password: '678912',
          id: 3,
        }
      ]
```

File: src/Register/index.test.js [hunk 28]
```diff
@@ -1,18 +1,91 @@
...
+  userEvent.type(passwordField, ' {backspace}');
+  expect(await screen.findByText('Password required.')).toBeInTheDocument();
+  userEvent.type(passwordField, 'password');
   expect(screen.getByDisplayValue('password')).toBeInTheDocument();
+
+  fireEvent.click(screen.getByRole('button'));
+  expect(await screen.findByText('Username taken.')).toBeInTheDocument();
+  userEvent.type(usernameField, '_');
+  fireEvent.click(screen.getByRole('button'));
+  expect(screen.queryByText('Username taken.')).not.toBeInTheDocument();
```
Comment (ahmed-gz) [inline src/Register/index.test.js]: Similar to what you did in the test for login, you could check if user was redirected to the home page here.

File: src/Register/index.test.js [hunk 29]
```diff
@@ -1,18 +1,91 @@
...
+  result: {
+    data: {
+      id: 1,
+    },
+  },
+};
+
+const mocks = [
+  mockedUsersResponse,
+  mockedUsersResponse,
```
Comment (ahmed-gz) [inline src/Register/index.test.js]: Same as the mock for login, this duplication doesn't feel like a realistic mock of real data, please see if adding data to `users` would work like my other similar suggestion

File: src/Login/index.test.js [hunk 30]
```diff
@@ -1,15 +1,75 @@
...
+
+  const passwordField = screen.getByLabelText('Password');
+  userEvent.type(passwordField, ' {backspace}');
+  expect(await screen.findByText('Password required.')).toBeInTheDocument();
+  userEvent.type(passwordField, 'password');
   expect(screen.getByDisplayValue('password')).toBeInTheDocument();
+
+  expect(loginButton).toBeEnabled();
+
+  fireEvent.click(loginButton);
```
Comment (ahmed-gz) [inline src/Login/index.test.js]: You can use `userEvent` here instead of `fireEvent`, and it's usually a best practice to always use `userEvent` whenever is possible since it's what's close to mimicking user behavior, check this article that explains the reasons: https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-testing-libraryuser-event

File: src/Login/index.test.js [hunk 31]
```diff
@@ -1,15 +1,75 @@
...
+  expect(await screen.findByText('Incorrect username or password.')).toBeInTheDocument();
+  expect(localStorage.getItem('user')).toBeNull();
+  userEvent.type(usernameField, '{backspace}');
+  fireEvent.click(loginButton);
```
Comment (ahmed-gz) [inline src/Login/index.test.js]: You can use `userEvent` here

File: src/Register/index.test.js [hunk 32]
```diff
@@ -1,18 +1,91 @@
...
   expect(screen.getByDisplayValue('Smith')).toBeInTheDocument();
-  userEvent.type(screen.getByLabelText('Password'), 'password');
+
+  const passwordField = screen.getByLabelText('Password');
+  userEvent.type(passwordField, ' {backspace}');
+  expect(await screen.findByText('Password required.')).toBeInTheDocument();
+  userEvent.type(passwordField, 'password');
   expect(screen.getByDisplayValue('password')).toBeInTheDocument();
+
+  fireEvent.click(screen.getByRole('button'));
```
Comment (ahmed-gz) [inline src/Register/index.test.js]: You can use `userEvent` here

File: src/Register/index.test.js [hunk 33]
```diff
@@ -1,18 +1,91 @@
...
+  expect(await screen.findByText('Username taken.')).toBeInTheDocument();
+  userEvent.type(usernameField, '_');
+  fireEvent.click(screen.getByRole('button'));
```
Comment (ahmed-gz) [inline src/Register/index.test.js]: You can use `userEvent` here

File: src/Login/index.test.js [hunk 34]
```diff
@@ -1,15 +1,75 @@
...
+  },
+};
+
+const mocks = [
+  mockedUsersResponse,
+  mockedUsersResponse,
+  mockedUsersResponse,
+];
+
+it('renders the login page', async () => {
```
Comment (ahmed-gz) [inline src/Login/index.test.js]: If it's not too much to ask, it would be great if you split the edge cases under this test into multiple ones that are small and have a specific goal, because this one covers a lot of cases which is good but makes it hard to read and will be even harder to maintain, for example you can split them like this and go even further with more tests if you want:
```
const App = () => (
  <MockedProvider mocks={mocks} addTypename={false}>
    <AuthProvider>
      <Router initialEntries={[LOGIN_PATH]}>
        <Switch>
          <Route exact path={HOME_PATH} component={() => <div>Home</div>} />
          <Route path={LOGIN_PATH} component={Login} />
        </Switch>
      </Router>
    </AuthProvider>
  </MockedProvider>
);

describe('Login page', () => {
  it('should render login form with initial state', async () => {
    render(<App />);
    // TODO
  });
  it('should not allow user to login when uername is invalid', async () => {
    render(<App />);
    // TODO
  });

  it('should not allow user to login when password is invalid', async () => {
    render(<App />);
    // TODO
  });

  it('should allow user to login when both uername and password are valid', async () => {
    render(<App />);
    // TODO
  });

  it('should {any other edge cases you can think of}', async () => {
    render(<App />);
    // TODO
  });
});
```

File: src/Register/index.test.js [hunk 35]
```diff
@@ -1,18 +1,91 @@
...
+  },
+};
+
+const mocks = [
+  mockedUsersResponse,
+  mockedUsersResponse,
+  mockedRegisterResponse,
+];
+
+it('renders the register page', async () => {
```
Comment (ahmed-gz) [inline src/Register/index.test.js]: Same as my suggestion on the login page, it would be great if you can break this test down into smaller ones.

File: src/graphql/queries.js [hunk 36]
```diff
@@ -0,0 +1,12 @@
+import { gql } from 'graphql-tag';
+
+const QUERY_USERS = gql`
+query QUERY_USERS {
+  users {
+    username
+    id
+    password
+  }
+}`;
```
Comment (ahmed-gz) [inline src/graphql/queries.js]: I think you're missing indentation here, not sure why prettier didn't suggest it for you:
```
const QUERY_USERS = gql`
  query QUERY_USERS {
    users {
      username
      id
      password
    }
  }
`;
```

File: src/graphql/mutations.js [hunk 37]
```diff
@@ -0,0 +1,21 @@
...
+  createUser(
+    username: $username
+    firstName: $firstName
+    lastName: $lastName
+    password: $password
+    userRoleId: 3
+  ) {
+    id
+  }
+}`;
```
Comment (ahmed-gz) [inline src/graphql/mutations.js]: I think you're missing indentation here, not sure why prettier didn't suggest it for you:
```
const REGISTER = gql`
  mutation REGISTER($username: String!, $firstName: String!, $lastName: String!, $password: String!) {
    createUser(username: $username, firstName: $firstName, lastName: $lastName, password: $password, userRoleId: 3) {
      id
    }
  }
`;
```

**Review by ahmed-gz — CHANGES_REQUESTED** [review]
Left more suggestions to make this PR better 😄

File: src/Login/index.test.js (hunk 27 above)
Comment (Aaron-Goldman) [reply src/Login/index.test.js]: The MockedProvider component requires separate mock response objects for each query, even if the query is being repeated.

Comment (Aaron-Goldman) [reply src/Login/index.test.js]: Though it's unnecessary anyway now after splitting up the tests into multiple cases.

File: src/Login/index.test.js [hunk 38]
```diff
@@ -8,68 +8,106 @@ import Login from './index';
...
+    userEvent.click(loginButton);
+    expect(await screen.findByText('Incorrect username or password.')).toBeInTheDocument();
+  });
 
-  const passwordField = screen.getByLabelText('Password');
-  userEvent.type(passwordField, ' {backspace}');
-  expect(await screen.findByText('Password required.')).toBeInTheDocument();
-  userEvent.type(passwordField, 'password');
-  expect(screen.getByDisplayValue('password')).toBeInTheDocument();
+  it('saves user state in local storage on failed login', () => {
```
Comment (ahmed-gz) [inline src/Login/index.test.js]: Shouldn't this be described as `doesn't save user state in local storage on failed login` or  `clears user state in local storage on failed login`?

File: src/Login/index.test.js [hunk 39]
```diff
@@ -8,68 +8,106 @@ import Login from './index';
...
+    setup();
+    expect(localStorage.getItem('user')).toBeNull();
```
Comment (ahmed-gz) [inline src/Login/index.test.js:110]: Shouldn't you try to mimic a user trying to login with `userEvent.click(loginButton)` before this assertion to make sure user state will never be saved in local storage when login fails. Because even if the typed values are correct, user state won't be saved in local storage unless user clicks on the button right?

File: src/Login/index.test.js [hunk 40]
```diff
@@ -8,68 +8,106 @@ import Login from './index';
...
+  });
 
-  expect(loginButton).toBeEnabled();
+  it('redirects to home path on successful login', async () => {
+    const { usernameField, passwordField, loginButton } = setup();
+    userEvent.type(usernameField, 'User');
+    userEvent.type(passwordField, 'password');
+    userEvent.click(loginButton);
+    expect(await screen.findByText('Home')).toBeInTheDocument();
+  });
```
Comment (ahmed-gz) [inline src/Login/index.test.js:129]: You could add a similar test for when user has incorrect credentials and expect not redirecting to home page,

File: src/Login/index.test.js [hunk 41]
```diff
@@ -8,68 +8,106 @@ import Login from './index';
...
+    expect(localStorage.getItem('user')).toBeNull();
+  });
 
-  expect(loginButton).toBeEnabled();
+  it('redirects to home path on successful login', async () => {
+    const { usernameField, passwordField, loginButton } = setup();
+    userEvent.type(usernameField, 'User');
+    userEvent.type(passwordField, 'password');
+    userEvent.click(loginButton);
+    expect(await screen.findByText('Home')).toBeInTheDocument();
```
Comment (ahmed-gz) [inline src/Login/index.test.js:127]: This looks good, but you can go one step further and check if the URL got changed to `HOME_PATH`, could you try if this would work? **expect(window.location.href).toBe(`${window.location.origin}${HOME_PATH}`)**

File: src/Login/index.test.js (hunk 27 above)
Comment (ahmed-gz) [reply src/Login/index.test.js]: @Aaron-Goldman but would there be an issue with having multiple users in the response rather than just one? My main goal is to mimic a real situation where we test that this code would be solid when we are searching for a user that's part of a large collections of existing users: 
`result.data.users.find((u) => u.username === username && u.password === password);`

But if we only have one user in the mock, our test won't be as efficient I think, does it make sense?

File: src/Login/index.test.js [hunk 42]
```diff
@@ -8,68 +8,106 @@ import Login from './index';
...
+    return { loginButton, usernameField, passwordField };
+  };
+
+  it('renders the login page', () => {
+    const { loginButton, usernameField, passwordField } = setup();
+    expect(screen.getByRole('form')).toBeInTheDocument();
+    expect(screen.getByText('Register Now')).toBeInTheDocument();
+    expect(loginButton).toBeInTheDocument();
+    expect(usernameField).toBeInTheDocument();
+    expect(passwordField).toBeInTheDocument();
```
Comment (ahmed-gz) [inline src/Login/index.test.js]: These assertions are redundant since we already do the same thing inside `setup()` with `getBy*`:
```
    expect(loginButton).toBeInTheDocument();
    expect(usernameField).toBeInTheDocument();
    expect(passwordField).toBeInTheDocument();
```

File: src/Register/index.test.js [hunk 43]
```diff
@@ -1,91 +1,136 @@
...
+  it('renders the register page', () => {
+    const {
+      usernameField, firstNameField, lastNameField, passwordField, registerButton,
+    } = setup();
+    expect(screen.getByRole('form')).toBeInTheDocument();
+    expect(usernameField).toBeInTheDocument();
+    expect(firstNameField).toBeInTheDocument();
+    expect(lastNameField).toBeInTheDocument();
+    expect(passwordField).toBeInTheDocument();
+    expect(registerButton).toBeInTheDocument();
```
Comment (ahmed-gz) [inline src/Register/index.test.js]: These assertions are redundant since we already do the same thing inside `setup()` with `getBy*`:
```
    expect(usernameField).toBeInTheDocument();
    expect(firstNameField).toBeInTheDocument();
    expect(lastNameField).toBeInTheDocument();
    expect(passwordField).toBeInTheDocument();
    expect(registerButton).toBeInTheDocument();
```

File: src/Register/index.test.js [hunk 44]
```diff
@@ -1,91 +1,136 @@
...
+  });
```
Comment (ahmed-gz) [inline src/Register/index.test.js:91]: I think you forgot to check for the state of `registerButton` : `expect(registerButton).toBeDisabled();`

**Review by ahmed-gz — CHANGES_REQUESTED** [review]
Looks really good, we are almost there, left small suggestions!

File: src/Login/index.test.js (hunk 39 above)
Comment (Aaron-Goldman) [reply src/Login/index.test.js:110]: Thanks, forgot to copy in the code for this.

File: src/Login/index.test.js [hunk 45]
```diff
@@ -8,68 +8,121 @@ import Login from './index';
...
-          password: 'password',
-          id: 1,
-        },
-      ],
+const mocks = [
+  {
+    request: { query: QUERY_USERS },
+    result: {
+      data: {
+        users: [
```
Comment (ahmed-gz) [inline src/Login/index.test.js:16]: @Aaron-Goldman Would there be an issue with having multiple users in here rather than just one? My main goal is to mimic a real situation where we test that this code would be solid when we are searching for a user that's part of a large collections of existing users: 
`result.data.users.find((u) => u.username === username && u.password === password);`

Ideally we want to mock having at least 3 users in the database like this:
```
      users: [
        {
          username: 'Danny',
          password: '632569',
          id: 1,
        },
        {
          username: 'James',
          password: '12345',
          id: 2,
        },
        {
          username: 'William',
          password: '678912',
          id: 3,
        }
      ]
```

But if we only have one user in the mock, our test won't be as efficient I think, does it make sense?

File: src/Register/index.test.js [hunk 46]
```diff
@@ -1,91 +1,148 @@
...
-          password: 'password',
-          id: 1,
-        },
-      ],
+const mocks = [
+  {
+    request: { query: QUERY_USERS },
+    result: {
+      data: {
+        users: [
```
Comment (ahmed-gz) [inline src/Register/index.test.js:17]: Same as above, it would be great if we can have multiple users in here or achieve that in a different way!

**Review by ahmed-gz — CHANGES_REQUESTED** [review]
Looks really good, once we have multiple users in the mock for login and register we can merge this PR 👍

**Review by ahmed-gz — APPROVED** [review]
Looks great, feel free to merge it, awesome job 👏
